src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagKey.cs

```diff
@@ -0,0 +1,428 @@
+                .Append(_key);
+
+            if (!string.IsNullOrEmpty(_varyBy))
+            {
+                builder
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(nameof(_varyBy))
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(_varyBy);
+            }
+
+            AddStringCollection(builder, nameof(_cookies), _cookies);
+            AddStringCollection(builder, nameof(_headers), _headers);
+            AddStringCollection(builder, nameof(_queries), _queries);
+            AddStringCollection(builder, nameof(_routeValues), _routeValues);
+
+            if (_varyByUser)
+            {
+                builder
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(nameof(_varyByUser))
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(_username);
+            }
+
+            return _generatedKey = builder.ToString();
+        }
+
+        /// <summary>
+        /// Creates a hashed value of the key.
+        /// </summary>
+        /// <returns>A cryptographic hash of the key.</returns>
+        public string GenerateHashedKey()
+        {
+            var key = GenerateKey();
+
+            // The key is typically too long to be useful, so we use a cryptographic hash
+            // as the actual key (better randomization and key distribution, so small vary
+            // values will generate dramatically different keys).
+            using (var sha = SHA256.Create())
+            {
+                var contentBytes = Encoding.UTF8.GetBytes(key);
+                var hashedBytes = sha.ComputeHash(contentBytes);
+                return Convert.ToBase64String(hashedBytes);
+            }
+        }
+
+        private static void AddStringCollection(
+            StringBuilder builder,
+            string collectionName,
+            IList<KeyValuePair<string, string>> values)
+        {
+            if (values == null || values.Count == 0)
+            {
+                return;
+            }
+
+            // keyName(param1=value1|param2=value2)
+            builder
+                .Append(CacheKeyTokenSeparator)
+                .Append(collectionName)
+                .Append("(");
+
+            for (var i = 0; i < values.Count; i++)
+            {
+                var item = values[i];
+
+                builder
+                    .Append(item.Key)
+                    .Append(CacheKeyTokenSeparator)
+                    .Append(item.Value)
+                    .Append(CacheKeyTokenSeparator);
+            }
+            
+            builder.Append(")");
+        }
+
+        protected static IList<string> Tokenize(string value)
+        {
+            var values = value.Split(AttributeSeparator, StringSplitOptions.RemoveEmptyEntries);
+            if (values.Length == 0)
+            {
+                return values;
+            }
+
+            var trimmedValues = new List<string>();
```

Comment:
Edit in place? Does this method have to return `IList<string>` or does `string[]` suffice?